trace-forward/src/Trace/Forward/Network/Forwarder.hs

```diff
@@ -0,0 +1,151 @@
+import           Ouroboros.Network.Util.ShowProxy (ShowProxy(..))
+
+import           Trace.Forward.Configuration (ForwarderConfiguration (..), HowToConnect (..))
+import           Trace.Forward.Queue (readItems)
+import qualified Trace.Forward.Protocol.Forwarder as Forwarder
+import qualified Trace.Forward.Protocol.Codec as Forwarder
+import           Trace.Forward.Protocol.Limits (byteLimitsTraceForward, timeLimitsTraceForward)
+
+connectToAcceptor
+  :: (CBOR.Serialise lo,
+      ShowProxy lo)
+  => ForwarderConfiguration lo
+  -> TBQueue lo
+  -> IO ()
+connectToAcceptor config@ForwarderConfiguration{..} loQueue = withIOManager $ \iocp ->
+  case acceptorEndpoint of
+    LocalPipe localPipe -> do
+      let snocket = localSnocket iocp localPipe
+          address = localAddressFromPath localPipe
+      doConnectToAcceptor snocket address noTimeLimitsHandshake app
+    RemoteSocket host port -> do
+      acceptorAddr:_ <- Socket.getAddrInfo Nothing (Just $ T.unpack host) (Just $ show port)
+      let snocket = socketSnocket iocp
+          address = Socket.addrAddress acceptorAddr
+      doConnectToAcceptor snocket address timeLimitsHandshake app
+ where
+  app =
+    OuroborosApplication $ \_connectionId _shouldStopSTM ->
+      [ MiniProtocol
+          { miniProtocolNum    = MiniProtocolNum 1
+          , miniProtocolLimits = MiniProtocolLimits { maximumIngressQueue = maxBound }
+          , miniProtocolRun    = forwardTraceObjects config loQueue
+          }
+      ]
+
+doConnectToAcceptor
+  :: Snocket IO fd addr
+  -> addr
+  -> ProtocolTimeLimits (Handshake UnversionedProtocol Term)
+  -> OuroborosApplication 'InitiatorMode addr LBS.ByteString IO () Void
+  -> IO ()
+doConnectToAcceptor snocket address timeLimits app =
+  connectToNode
+    snocket
+    unversionedHandshakeCodec
+    timeLimits
+    (cborTermVersionDataCodec unversionedProtocolDataCodec)
+    nullNetworkConnectTracers
+    acceptableVersion
+    (simpleSingletonVersions
+       UnversionedProtocol
+       UnversionedProtocolData
+       app)
+    Nothing
+    address
+
+forwardTraceObjects
+  :: (CBOR.Serialise lo,
+      ShowProxy lo)
+  => ForwarderConfiguration lo
+  -> TBQueue lo
+  -> RunMiniProtocol 'InitiatorMode LBS.ByteString IO () Void
+forwardTraceObjects config loQueue =
+  InitiatorProtocolOnly $
+    MuxPeerRaw $ \channel -> do
+      cv <- newEmptyTMVarIO
+      siblingVar <- newTVarIO 2
+      (r, trailing) <-
+        runPeerWithLimits
+          (forwarderTracer config)
+          (Forwarder.codecTraceForward CBOR.encode CBOR.decode
+                                       CBOR.encode CBOR.decode
+                                       CBOR.encode CBOR.decode)
+          (byteLimitsTraceForward (fromIntegral . LBS.length))
+          timeLimitsTraceForward
+          channel
+          (Forwarder.traceForwarderPeer $ readItems config loQueue)
+      atomically $ putTMVar cv r
+      waitSibling siblingVar
```

Comment:
I think this just blocks forever.  The only place `siblingVar` is subtracted is in `waitSibling` which is called once for the given `siblingVar` created in the same monadic context.

What is the purpose of `siblingVar`?

Reply:
Honestly, I don't know. I just copied this snippet from `ouroboros-network-framework/test/Test/Ouroboros/Network/Subscription.hs`.

Reply:
But there we have two threads writing updating the `siblingVar`, so it has a chance to terminate.

Reply:
I removed it.